paymentsheet/src/androidTest/java/com/stripe/android/paymentsheet/utils/ActivityLaunchObserver.kt

```diff
@@ -0,0 +1,61 @@
+package com.stripe.android.paymentsheet.utils
+
+import android.app.Activity
+import android.app.Application
+import android.os.Bundle
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+
+internal class ActivityLaunchObserver(
+    private val expectedActivityType: Class<out Activity>
+) {
+    @Volatile
+    private var unregisterer: (() -> Unit) = { }
+
+    private val launchedCountDownLatch = CountDownLatch(1)
+
+    fun prepareForLaunch(host: Activity) {
+        val activityLifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
+            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
+            }
+
+            override fun onActivityStarted(activity: Activity) {
+            }
+
+            override fun onActivityResumed(activity: Activity) {
+                if (expectedActivityType.isInstance(activity)) {
+                    launchedCountDownLatch.countDown()
+                }
+            }
+
+            override fun onActivityPaused(activity: Activity) {
+            }
+
+            override fun onActivityStopped(activity: Activity) {
+            }
+
+            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
+            }
+
+            override fun onActivityDestroyed(activity: Activity) {
+            }
+
+        }
+        val application = host.applicationContext as Application
+        application.registerActivityLifecycleCallbacks(activityLifecycleCallbacks)
+        unregisterer = {
+            application.unregisterActivityLifecycleCallbacks(activityLifecycleCallbacks)
+        }
+    }
+
+    fun awaitLaunch() {
+        try {
+            if (!launchedCountDownLatch.await(5, TimeUnit.SECONDS)) {
```

Comment:
does this stop the rest of the test from running? Should it? I can definitely see myself checking for this log line to determine if a test flaked for this reason, but I wonder if it should also just prevent the test from continuing to run

Reply:
It doesn't. But it seems reasonable to just throw instead of log here. That way we get the message and the execution of the rest of the test will stop.